### PR TITLE
greg/NFSE-5159-edge-map

### PR DESCRIPTION
**Pull request by beckerg**
Signed-off-by: Greg Becker <gbecker@micron.com>

### CODE REVIEW

File: lib/kvdb/ikvdb.c [hunk 1]
```diff
@@ -1882,32 +1882,54 @@ kvdb_kvs_set_ikvs(struct kvdb_kvs *kk, struct ikvs *ikvs)
...
 
-    merr_t             err = 0;
-    struct ikvdb_impl *self = ikvdb_h2r(handle);
-    struct kvdb_kvs *  kvs = NULL;
-    int                idx;
-
+    self = ikvdb_h2r(handle);
     if (self->ikdb_read_only)
-        goto out_immediate;
+        return 0;
```
Comment (alexttx) [inline lib/kvdb/ikvdb.c:1898]: return error here?

File: lib/cn/spill.c [hunk 2]
```diff
@@ -284,8 +284,21 @@ kv_spill(struct cn_compaction_work *w)
 
     hash = pfx_obj_hash64(&curr.kobj, hashlen);
 
-    if (w->cw_outc > 1) {
-        cnum = cn_tree_route_create(w->cw_tree, hash, w->cw_level);
+    /* Check w->cw_tree because merge_test sets it to NULL.
```
[the remark this answers is not in the document]
Comment (beckerg) [reply lib/cn/spill.c:287]: That won't work, I need to deref tree in this function.
I suppose I could move this entire block into cn_tree_route_create(), which means I have to pass the kobj...

File: lib/kvdb/ikvdb.c [hunk 3]
```diff
@@ -1882,32 +1882,54 @@ kvdb_kvs_set_ikvs(struct kvdb_kvs *kk, struct ikvs *ikvs)
...
-        goto out_immediate;
-    }
+    if (ev(!kvs))
+        return merr(ENOMEM);
 
     strlcpy(kvs->kk_name, kvs_name, sizeof(kvs->kk_name));
 
+    /* The following block is transitional code to assist
+     * with initializing the hard-coded route map.
+     */
```
Comment (alexttx) [inline lib/kvdb/ikvdb.c:1912]: Tag with HSE_REVISIT?

File: lib/include/hse_ikvdb/ikvdb.h [hunk 4]
```diff
@@ -364,7 +364,7 @@ ikvdb_kvs_count(struct ikvdb *kvdb, unsigned int *count);
  * @params:    KVS cparams
  */
 merr_t
-ikvdb_kvs_create(struct ikvdb *kvdb, const char *kvs_name, const struct kvs_cparams *params);
+ikvdb_kvs_create(struct ikvdb *kvdb, const char *kvs_name, struct kvs_cparams *params);
```
Comment (alexttx) [inline lib/include/hse_ikvdb/ikvdb.h:367]: HSE_REVISIT: convert last arg back to const

File: lib/cn/spill.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexttx) [reply lib/cn/spill.c:287]: testing is hard...

File: lib/kvdb/ikvdb.c (hunk 1 above)
Comment (beckerg) [reply lib/kvdb/ikvdb.c:1898]: It didn't used to return an error, so I don't want to break it..  Dunno..

File: lib/cn/route.h [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+struct kvs_cparams;
+
+struct route_map *
+route_map_create(const struct kvs_cparams *cp, const char *kvsname);
+
+void
+route_map_destroy(struct route_map *map);
+
+uint
+route_map_lookup(struct route_map *map, const void *pfx, uint pfxlen);
```
Comment (alexttx) [inline lib/cn/route.h:16]: #ifdef guards?

File: lib/cn/route.h [hunk 6]
```diff
@@ -0,0 +1,16 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (C) 2022 Micron Technology, Inc.  All rights reserved.
+ */
+
+struct route_map;
+struct kvs_cparams;
+
+struct route_map *
+route_map_create(const struct kvs_cparams *cp, const char *kvsname);
```
Comment (alexttx) [inline lib/cn/route.h:10]: I like `edge_map` better than `route_map`...

File: lib/kvdb/ikvdb.c (hunk 3 above)
Comment (beckerg) [reply lib/kvdb/ikvdb.c:1912]: done

File: lib/include/hse_ikvdb/ikvdb.h (hunk 4 above)
Comment (beckerg) [reply lib/include/hse_ikvdb/ikvdb.h:367]: done

File: lib/cn/route.h (hunk 5 above)
Comment (beckerg) [reply lib/cn/route.h:16]: Doh!  Done.

File: lib/cn/route.h (hunk 6 above)
Comment (beckerg) [reply lib/cn/route.h:10]: i can change it..

Comment (alexttx) [reply lib/cn/route.h:10]: or maybe route is better.  i dunno.  i'm on the edge.

File: lib/kvdb/ikvdb.c (hunk 1 above)
Comment (alexttx) [reply lib/kvdb/ikvdb.c:1898]: bizarre

**Comment by beckerg** [comment]
Lemme revert and rename route to edge.

**Comment by alexttx** [comment]
Impressive to be able to convert to fixed-fanout edge map with so few code changes!

**Review by alexttx — APPROVED** [review]

File: lib/cn/spill.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (beckerg) [reply lib/cn/spill.c:287]: aint it though?  lemme try to move this block into cn_tree_route_create()...

File: lib/cn/route.h (hunk 6 above)
Comment (beckerg) [reply lib/cn/route.h:13]: Very punny!   Easier to change now before everyone starts writing to it, I'm good either way..

**Comment by beckerg** [comment]
Just now squashed to make it easier to review..

File: lib/cn/route.c [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    while (*link) {
+        struct route_node *this;
+        int rc;
+
+        this = rb_entry(*link, struct route_node, rtn_node);
+
+        rc = memcmp(key, this->rtn_keybuf, len);
+
+        if (rc < 0) {
+            dir = -1;
```
Comment (nabeelmmd) [inline lib/cn/route.c:89]: "dir" is used to keep track of the inorder successor of the searched key, right?

Comment (beckerg) [reply lib/cn/route.c:89]: Yes, I need to keep track of whether I've ever gone left in order to choose the correct node for cases where we don't have an exact edge key match and hence fall out the bottom of the loop.
Was maddening trying to figure this out.  Someone should review it carefully!!!

**Comment by beckerg** [comment]
See ~gbecker/routemap for some examples.

File: lib/cn/route.c (hunk 7 above)
Comment (nabeelmmd) [reply lib/cn/route.c:89]: This makes sense to me..
The last node from which a left turn was taken in the traversal path is the inorder successor of the searched key.
If there was no left turn taken in the traversal path, then the inorder successor is the last parent node.

File: lib/cn/route.c [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+    }
+
+
+    return map;
+}
+
+void
+route_map_destroy(struct route_map *map)
+{
+    if (!map)
```
Comment (nabeelmmd) [inline lib/cn/route.c:243]: This check can be removed as free(NULL) should be fine..

**Review by nabeelmmd — APPROVED** [review]

File: lib/cn/route.c [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+    }
+
+    free(fmt);
+
+    if (!rb_first(&map->rtm_root)) {
+        free(map);
+        return NULL;
+    }
+
+    if (1) {
```
Comment (alexttx) [inline lib/cn/route.c:217]: make up your mind :)

File: lib/cn/route.c [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+    n = sscanf(buf, "%u%u%ms%u", &fanout, &pfxlen, &fmt, &skip);
+
+    if (n < 3 || fanout != cp->fanout || pfxlen != cp->pfx_len || skip < 1) {
+        log_err("fanout (%u vs %u), pfxlen (%u vs %u), skip %u",
+                fanout, cp->fanout, pfxlen, cp->pfx_len, skip);
+        return NULL;
+    }
+
+    if (n < 4 || skip < 1)
+        skip = 1;
```
[the remark this answers is not in the document]
Comment (beckerg) [reply lib/cn/route.c:146]: Right.  I think I used to initialize skip earlier and also it was a more recent addition to the log message.
Nabeel can fix it when he adds his comments :-)

File: lib/cn/route.c (hunk 9 above)
Comment (beckerg) [reply lib/cn/route.c:217]: We'll whack it eventually...

File: lib/cn/route.c (hunk 8 above)
Comment (beckerg) [reply lib/cn/route.c:243]: Yep.  I used to do the "dump the edge keys" here, but then I moved it into create...

File: lib/kvdb/ikvdb.c (hunk 1 above)
Comment (tristan957) [reply lib/kvdb/ikvdb.c:1898]: we should return an error here for 3.0

**Pull request merged**